bls12_381/src/fp.rs

```diff
@@ -545,6 +608,181 @@ impl Fp {
     }
 }
 
+pub struct FpRepr([u8; 48]);
+
+impl fmt::Debug for FpRepr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "0x")?;
+        for &b in self.0.iter() {
+            write!(f, "{:02x}", b)?;
+        }
+        Ok(())
+    }
+}
+
+impl Default for FpRepr {
+    fn default() -> Self {
+        FpRepr([0; 48])
+    }
+}
+
+impl AsRef<[u8]> for FpRepr {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[..]
+    }
+}
+
+impl AsMut<[u8]> for FpRepr {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0[..]
+    }
+}
+
+impl From<Fp> for FpRepr {
+    fn from(value: Fp) -> FpRepr {
+        value.into_repr()
+    }
+}
+
+impl<'a> From<&'a Fp> for FpRepr {
+    fn from(value: &'a Fp) -> FpRepr {
+        value.into_repr()
+    }
+}
+
+impl From<u64> for Fp {
+    fn from(val: u64) -> Fp {
+        Fp([val, 0, 0, 0, 0, 0]) * R2
+    }
+}
+
+impl BitAnd<u64> for Fp {
```

Comment:
These (this and the Shr) are weird. I don't think we should support them as part of the `Field` trait's contract. If someone needs to do this stuff they should deal with the existing API somehow.

Reply:
They are part of the `PrimeField` trait contract, due to `bellman::multiexp` deriving buckets, and `zcash_primitives::pedersen_hash` generating a window table. This is one of the constraints I was contemplating moving to a `ScalarField` trait, so that we could implement these only for `bls12_381::Scalar` and `jubjub::Fr`, not `bls12_381::Fp`.

Reply:
See 08500ee71275e7b6127e7b512c0763b058acf714 and 49f119fb033dac88165028b7117e3d23e9f469a8 (the `bellman::multiexp` change was accidentally left in the latter).

Reply:
Yeah, those two examples are niche. I think we can implement this inline and not provide it as part of the crate. I also think the implementation of multiexp is weird and shouldn't need to use those operations... well, probably anyway.